packages/google_sign_in/lib/src/google_sign_in.dart

```diff
@@ -0,0 +1,302 @@
+
+import 'device_flow_widget.dart';
+import 'utils.dart' as utils;
+
+/// The class that represents a Google user account.
+class GoogleUser {
+  /// Creates an instance of [GoogleUser].
+  GoogleUser({
+    required this.userId,
+    required this.profile,
+    required this.authentication,
+    List<String>? scope,
+  }) : grantedScope = scope ?? <String>[];
+
+  /// The Google user ID.
+  final String userId;
+
+  /// The basic profile data for the user.
+  final ProfileData profile;
+
+  /// The authentication info for the user.
+  final Authentication authentication;
+
+  /// The API scopes granted to the app.
+  final List<String> grantedScope;
+}
+
+/// The class that holds parameters for OAuth request.
+class Configuration {
+  /// Creates an instance of [Configuration].
+  const Configuration({
+    required this.clientId,
+    required this.clientSecret,
+    this.scope = const <String>[],
+  });
+
+  /// The unique public identifier for apps that is issued by the authorization
+  /// server. It's analogous to a login id.
+  final String clientId;
+
+  /// The secret credential knwon only to the application and the authorization
+  /// server. It's analogous to a password.
+  final String clientSecret;
+
+  /// The amount of resources to access on behalf of the user.
+  final List<String> scope;
+}
+
+/// The class that represents OAuth 2.0 entities after sign-in.
+class Authentication {
+  /// Creates an instance of [Authentication].
+  const Authentication({
+    required this.clientId,
+    required this.accessToken,
+    required this.accessTokenExpirationDate,
+    this.refreshToken,
+    this.idToken,
+    this.idTokenExpirationDate,
+  });
+
+  /// The client ID associated with the authentication.
+  final String clientId;
+
+  /// The OAuth2 access token to access Google services.
+  final String accessToken;
+
+  /// The estimated expiration date of [accessToken].
+  final DateTime accessTokenExpirationDate;
+
+  /// The OAuth2 refresh token to exchange for new access tokens.
+  final String? refreshToken;
+
+  /// The OpenID Connect ID token that identifies the user.
+  final String? idToken;
+
+  /// The estimated expiration date of [idToken].
+  final DateTime? idTokenExpirationDate;
+}
+
+/// The class that represents basic profile information.
+class ProfileData {
+  /// Creates an instance of [ProfileData].
+  const ProfileData({
+    required this.email,
+    required this.name,
+    this.givenName,
+    this.familyName,
+    this.picture,
+  });
+
+  /// The Google user's email.
+  final String email;
+
+  /// The Google user's full name.
+  final String name;
+
+  /// The Google user's given name.
+  final String? givenName;
+
+  /// The Google user's family name.
+  final String? familyName;
+
+  /// The Uri of Google user's profile picture.
+  final Uri? picture;
+
+  /// Creates a [ProfileData] from json object.
+  static ProfileData fromJson(Map<String, dynamic> json) {
```

Comment:
I agree that `checkFormat` is somewhat too much. During initial design, I wanted error messages to specifically complain which value was unable for type casting in case server has loosely implemented OAuth 2.0 response spec (I was experiencing some cases during testing though it wasn't about response types).

```dart
json['verification_url'] as int;
// error message is:
//   flutter: type 'String' is not a subtype of type 'int' in type cast

utils.checkFormat<int>(<String>['verification_url'], json);
// error message is:
//   flutter: FormatException: parameter "verification_url" is not a int, is "https://www.google.com/device".
```

I'll remove it if reducing code is better in this case.


> The extra variables should not be necessary. You can write just json['given_name'] as String?.

Much cleaner syntax, thanks!

Reply:
> I'll remove it if reducing code is better in this case.

The check could be removed, or might have a better design. It is not clear from the signature (`utils.checkFormat<T>(keys, json)`) what the function does.

Reply:
Okay, I removed the function.